### ISSUE TITLE
[NativeAOT] linux-arm bring up

### ISSUE THREAD

**Issue by filipnavara** [dotnet/runtime#97729]
This is tracking issue for the known problems that need to be resolved to get working NativeAOT support on linux-arm platform.

Known issues:
- [ ] Inline TLS access emits incorrect code for Optimized=true which results in stack overflow
- [ ] Some code paths around Align8 / `FEATURE_64BIT_ALIGNMENT` may be unhandled (ref: https://github.com/dotnet/runtime/pull/97269#issuecomment-1909219996)
- [ ] Interlocked JIT tests fail

Other things requiring clean up:
- [ ] NativeAOT build integration support for linux-musl and linux-bionic on arm32
- [ ] Run the smoke tests to prevent regressions

**Comment by filipnavara** [comment]
State as of ba8993fa8c80a663dadd495db17ac9593bdf703b + PRs #97746, #97756 and #97757:
- Smoke tests in Debug build run and pass on Raspberry Pi 4, Raspberry Pi 5 and QEMU
- All library tests build (full checked build of `./build.sh clr.aot+libs+libs.tests -rc Checked -lc Release /p:TestNativeAot=true`  took 03:48:16.60 on Raspberry Pi 5)
- Library tests in Debug builds seem to run to some extent (no fail/pass numbers to report but multiple tests run to the end; many tests failed but they were supposed to be excluded on given platform and the runner run them anyway)
- Library tests in Release builds crash with various errors almost instantly

**Comment by filipnavara** [comment]
I run the smoke tests in Release configuration. Some of them reliably fail which makes the debugging easier. Apparently we now get incorrect answer for `InWriteBarrierHelper` in the SIGSEGV exception handler. I'll debug it later this week.

**Comment by NCLnclNCL** [comment]
When support window x86(32 bit) ???

**Comment by michaldobrodenka** [comment]
> When support window x86(32 bit) ???
Why?
win-x86 is not suported in Windows 11, only 10 which will be supported only for about a year.
Maybe some win iot?

**Comment by filipnavara** [comment]
> When support window x86(32 bit) ???

Please keep this issue on topic. I am doing this on my own and in my free time, I do not plan to work on win-x86 port. There's already an [open issue for that](https://github.com/dotnet/runtime/issues/86573).

**Comment by NCLnclNCL** [comment]
> > When support window x86(32 bit) ???
> > Why?
> > win-x86 is not suported in Windows 11, only 10 which will be supported only for about a year.
> > Maybe some win iot?

Window 64 bit can run 32 bit application and i need it to run it in old device or hooking for application 32 bit

**Comment by filipnavara** [comment]
So, for the last crash in GC suspension I may need some help with verifying some assumptions. I can easily reproduce it and it's happening at the same point in the same function:
- One thread gets interrupted at `System.Collections.Concurrent.ConcurrentUnifierW`2_Container<System.Reflection.Runtime.TypeInfos.NativeFormat.NativeFormatRuntimeNamedTypeInfo_UnificationKey__System___Canon>__TryGetValue + 86`.
- GC decoding tells that the location is a safepoint (which may or may not be true).
- At that point in code the scratch register R12 is used (`0xd92742 <+80>: ldr.w  r12, [lr, #0xc]`). Are scratch registers supposed to be used at safepoints? If yes, is there any limit on which ones? We currently don't pass them to the stack frame iterator from the context. Should we?

Decoding the GCInfo on the GC thread indeed shows that there's a live variable in register R12, and since it's a scratch register `regDisplay->pR12 == NULL`, which in turn causes the crash.

**Comment by jkotas** [comment]
Safe points should be only created for call returns. Scratch registers cannot be live at call returns. It is why they are not handled for safe point. (@VSadov is changing some of these invariants in #95565.)

What does the code around the safe point look like? It may be useful to generate JIT dump for the method in question to see why the JIT decided to emit the safe point at this spot.

**Comment by filipnavara** [comment]
> What does the code around the safe point look like? It may be useful to generate JIT dump for the method in question to see why the JIT decided to emit the safe point at this spot.

It's this code (the offsets are one-off, ie. +85 is really +86):
```
    0xd92730 <+63>: bhs    0x3227d2                  ; <+225> at ConcurrentUnifierW.cs:199
    0xd92732 <+65>: mov.w  lr, #0x18
    0xd92736 <+69>: mul    lr, r3, lr
    0xd9273a <+73>: add.w  lr, lr, #0x8
    0xd9273e <+77>: add    r1, lr
    0xd92740 <+79>: mov    lr, r1
    0xd92742 <+81>: ldr.w  r12, [lr, #0xc]
    0xd92746 <+85>: ldr.w  lr, [lr, #0x10]
-- SAFEPOINT HERE? --
    0xd9274a <+89>: ldr    r4, [r0, #0x4]
    0xd9274c <+91>: cmp    r4, lr
    0xd9274e <+93>: bne    0x32275a                  ; <+105> at ConcurrentUnifierW.cs:195
    0xd92750 <+95>:  ldr    r0, [r0]
    0xd92752 <+97>:  ldrsb.w lr, [r0]
    0xd92756 <+101>: cmp    r0, r12
    0xd92758 <+103>: beq    0x322770                  ; <+127> at ConcurrentUnifierW.cs:199
    0xd9275a <+105>: ldr    r3, [r1, #0x8]
    0xd9275c <+107>: cmp.w  r3, #0xffffffff
    0xd92760 <+111>: bne    0x322726                  ; <+53> at ConcurrentUnifierW.cs:185
    0xd92762 <+113>: movs   r0, #0x0
    0xd92764 <+115>: ldr    r4, [sp, #0x24]
    0xd92766 <+117>: str    r0, [r4]
    0xd92768 <+119>: pop.w  {r4, r5, r6, r11, lr}
    0xd9276c <+123>: add    sp, #0xc
```

My suspicion is that the `IsSafepoint` answer is already wrong. I can dump ten GCInfo to verify what exactly is in there but I wanted to be sure that's the right direction to look into.

**Comment by VSadov** [comment]
> One thread gets interrupted at

Right now threads can only be interrupted in interruptible code. Volatile registers can contain GC refs there.

Also threads can self-interrupt when hitting a hijacked return.Volatile regs are dead, but return registers may contain live GC regs.

After that it is unwinding that through returns when returns did not happen yet. Volatile registers are dead.

Since there are no calls around the interruption location in you sample, it is must be in fully interruptible method.

**Comment by filipnavara** [comment]
The crash happens only when the C runtime part is built with optimizations, so most likely there's an issue with decoding the GC info (the compiler is very eager to optimize out alignments when wrong pointer type is used). I'll dump the GC info.

> Since there are no calls around the interruption location in you sample, it is must be in fully interruptible method.

I really hope it's just misdecoded GC info... because a fully interruptible method should not use the R12 register (or we would need to save it from the frame which is trivial).

**Comment by VSadov** [comment]
The part that C optimizations matter is suspicious indeed.

For the r12 register i do not recall if its use for scratch is forbidden (can’t easily check that right now).  
Typically a register set for a leaf frame includes scratch registers, since leaf frames report them to gc.
If we are not on a call return, we are probably in a leaf.

**Comment by jkotas** [comment]
> a fully interruptible method should not use the R12 register (or we would need to save it from the frame which is trivial).

It is fine for fully interruptible methods to use R12 register to store GC reference. I think it is a bug that it is not initialized in the REGDISPLAY in `StackFrameIterator::InternalInit(Thread * pThreadToWalk, NATIVE_CONTEXT* pCtx, uint32_t dwFlags)`.

**Comment by filipnavara** [comment]
Thanks. I came to the same conclusion. The GCInfo shows that it's fully interruptible method.

I'll send a PR.

**Comment by filipnavara** [comment]
State as of df0778dc9eb9f15c9270ba1a09d475253018e824:
- NativeAOT smoke tests run and pass both in Debug and Release builds
- Running System.Runtime.Tests in Release builds is blocked by https://github.com/dotnet/runtime/issues/97912 since Xunit does reflection of the metadata which contains float/double values

**Comment by filipnavara** [comment]
With #97917 and #97919 the System.Runtime.Tests get quite far:
```
  [FAIL] System.Tests.TimeSpanTests.Division(timeSpan: 366.00:00:00, factor: -2.7182818284590451, expected: -994.21:23:15.2922633)
  Assert.Equal() Failure: Values are not within 14 decimal places
  Expected: -2.71828182845905 (rounded from -2.7182818284590451)
  Actual:   -2.7182818284590402 (rounded from -2.7182818284590446)
     at System.Tests.TimeSpanTests.Division(TimeSpan timeSpan, Double factor, TimeSpan expected) + 0x39
     at System.Runtime!<BaseAddress>+0x23bcff0
     at System.Reflection.DynamicInvokeInfo.InvokeWithFewArguments(IntPtr, Byte&, Byte&, Object[], BinderBundle, Boolean) + 0x55
  [FAIL] System.Tests.TimeSpanTests.NamedDivision(timeSpan: 366.00:00:00, factor: -2.7182818284590451, expected: -994.21:23:15.2922633)
  Assert.Equal() Failure: Values are not within 14 decimal places
  Expected: -2.71828182845905 (rounded from -2.7182818284590451)
  Actual:   -2.7182818284590402 (rounded from -2.7182818284590446)
     at System.Tests.TimeSpanTests.NamedDivision(TimeSpan timeSpan, Double factor, TimeSpan expected) + 0x39
     at System.Runtime!<BaseAddress>+0x23bcff0
     at System.Reflection.DynamicInvokeInfo.InvokeWithFewArguments(IntPtr, Byte&, Byte&, Object[], BinderBundle, Boolean) + 0x55
  Finished System.Runtime.Tests, Version=9.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51

  Tests run: 63505, Errors: 0, Failures: 2, Skipped: 127. Time: 30.8337846s
```

There may still be some flaky errors.

**Comment by filipnavara** [comment]
With #97964 we pass the NativeAOT smoke tests and System.Runtime.Tests.

**Comment by filipnavara** [comment]
I am slowly running out of things to fix. Would it make sense to add CI pipeline to prevent build breaks? If so, which one would be the right one? There are some existing pipelines for smoke test, runtime-extra-platforms, nativeaot-outerloop, and runtime-community. Build systems are really my strong area but it would be nice to have this. There's at least one open PR that would break the build due to missing platform check, and if we can avoid these breaks (semi-)automatically it would be nice.

**Comment by MichalStrehovsky** [comment]
nativeaot-outerloop would work. I can have a look at that later today if you'd rather avoid the yaml.

I think we can set up official packaging as well. If there are no crazy problems we cannot solve, we can then bump linux-arm and linux-musl-arm into officially supported category in .NET 9.

**Comment by filipnavara** [comment]
> I can have a look at that later today if you'd rather avoid the yaml.

I would definitely appreciate help on that.

**Comment by filipnavara** [comment]
Remaining runtime tests that need fixing:
- [ ] JIT/Directed/Directed_1
- [ ] Loader/classloader/Casting/punning
- [ ] nativeaot/SmokeTests/Determinism
- [ ] JIT/Regression/JitBlue/GitHub_15291
- [ ] JIT/Regression/JitBlue/GitHub_15949

Notes:
- `nativeaot/SmokeTests/Determinism` used to pass in Debug and Release, so perhaps it's flaky or broken only in Checked builds
- `Loader/classloader/Casting/punning` passes in Debug but fails in Checked. The JIT generates code that directly tries to invoke fat pointer instead of checking the marker bit.

**Comment by filipnavara** [comment]
The failure in ` JIT/Directed/Directed_1` is instance of issue #95517. JIT generates the following code:
```
boxunboxvaluetype_ro`boxunboxvaluetype_ro_NullableTest43__BoxUnboxToNQGen<System.Nullable`1<boxunboxvaluetype_ro_WithMultipleGCHandleStruct>>:
    0x73bb60 <+0>:   push   {r0, r1, r2, r3}
    0x73bb62 <+2>:   push.w {r11, lr}
    0x73bb66 <+6>:   sub    sp, #0x58
    0x73bb68 <+8>:   add.w  r11, sp, #0x58
    0x73bb6c <+12>:  add    r1, sp, #0x60
    0x73bb6e <+14>:  movw   r0, #0xd5ee
    0x73bb72 <+18>:  movt   r0, #0xa
    0x73bb76 <+22>:  add    r0, pc
    0x73bb78 <+24>:  bl     0x64dfe0                  ; RhBox
    0x73bb7c <+28>:  adds   r1, r0, #0x4
    0x73bb7e <+30>:  add    r0, sp, #0x2c
    0x73bb80 <+32>:  movs   r2, #0x14
->  0x73bb82 <+34>:  blx    0x750d20                  ; symbol stub for: memcpy
```
The unboxing calls `memcpy` with a null value which causes a segfault instead of the expected null reference exception.

**Comment by emmauss** [comment]
Any plans on linux-bionic-arm builds being enabled for nuget?

**Comment by filipnavara** [comment]
> Any plans on linux-bionic-arm builds being enabled for nuget?

Yes. The runtime bits already work on Bionic. There is, however, some packaging issue but I would need @MichalStrehovsky to chime in on that.

**Comment by sonatique** [comment]
Hello all, thanks a lot for all these wonderful efforts. I really look forward to being able to use Arm32 AOT.
I am quite new here (I mean to the cutting edge .Net development process).
Given that all PR I can see above are merged into dotnet:main for a while now, I naively tried to use one of the latest dailybuild from here: https://github.com/dotnet/installer#installers-and-binaries: I unziped dotnet-sdk-8.0.300-preview.24209.26-win-x64.zip and started using it.
I was able to publish an Hello World as Win-x64 AOT, but trying `-c linux-arm` always tell me: "`error NETSDK1203: Ahead-of-time compilation is not supported for the target runtime identifier 'linux-arm'`" like if I was on plain .Net 8.

I think I tried to follow all indications found on https://github.com/dotnet/runtime/blob/main/docs/project/dogfooding.md and https://github.com/dotnet/installer?tab=readme-ov-file#installers-and-binaries but to no avail.

Could you tell me what I am doing wrong? Is there a resource to help newcomers to built AOT for linux ARM32? Sorry if it is obvious where to look at and I missed it.

Thanks in advance!

**Comment by Suchiman** [comment]
> I unziped dotnet-sdk-8.0.300-preview.24209.26-win-x64.zip and started using it

That is a .NET 8 build, you need a .NET 9 build, so this column 
![image](https://github.com/dotnet/runtime/assets/7110884/6dd137cc-b965-4a82-bea0-5c47d01ce2fe)

Also cross compilation is not supported out of the box yet, so you need to be on a linux machine to publish `linux-arm`.

**Comment by sonatique** [comment]
@Suchiman : thanks a lot for your crazy fast reply!
Ok I see, I completely overlooked the fact that the feature was targeted for .Net 9 (and obviously that cross-compilation is not yet supported).

OK cool, thanks again I'll try based on this!

**Comment by sonatique** [comment]
> Also cross compilation is not supported out of the box yet, so you need to be on a linux machine to publish `linux-arm`.

Sorry: just to be sure: do I need to be on a linux ARM32 machine or any linux will do? What about being on a linux x64 for instance? Does the processor architecture part of the cross-compilation works (unlike the OS part?) or there is no cross-compilation at all?

Thanks!

**Comment by filipnavara** [comment]
> do I need to be on a linux ARM32 machine or any linux will do? What about being on a linux x64 for instance? Does the processor architecture part of the cross-compilation works (unlike the OS part?) or there is no cross-compilation at all?

Any Linux will do. Compiling on linux-x64 (cross) and linux-arm (native) are the two tested scenarios.

**Comment by sonatique** [comment]
@filipnavara thanks for your equally prompt reply!

**Comment by sonatique** [comment]
Anyone as a (good) experience using WSL 2 for the linux-x64 machine when doing AOT for linux-arm?

**Comment by filipnavara** [comment]
> Anyone as a (good) experience using WSL 2 for the linux-x64 machine when doing AOT for linux-arm?

Yes, I used WSL 2 for the initial bring-up and even used QEMU user emulation to run some of the tests.

**Comment by filipnavara** [comment]
I am not a Docker fan but it may already exist at https://github.com/dotnet/dotnet-buildtools-prereqs-docker (and the prebuilt images from that repository). (/cc @am11 if you have more insight on that)

**Comment by sonatique** [comment]
Thanks a lot @filipnavara . I am not really into Docker myself neither, but it came to my mind that in this case it could be a good way to quickly have my first AOT built binary to play with (I am especially interested in comparing startup time and other performance related things in between AOT and non-AOT in order to validate whether AOT is the way to go for my project) before  getting serious about it.

Thanks again: I will try this images and then WSL 2 (or the other way round, whichever is simpler for me ;-) )

**Comment by am11** [comment]
On Apple Silicon, Docker depends on qemu and provides Roestta2 emulation option as an alternative. Unfortunately, Rosetta2 only supports x64 and arm64, so we have to use qemu which comes bundled with Docker Desktop. I was running into https://github.com/docker/for-mac/issues/7172, which turned out to be mmap issue. So `DOTNET_EnableWriteXorExecute=0` for the rescue!

@filipnavara, while preparing a solution for @sonatique, I ran into an assertion. 🥲 

Repro:
```Dockerfile
FROM --platform=linux/arm/v7 ubuntu:latest AS builder

RUN apt update && apt install -y clang zlib1g-dev curl

RUN mkdir -p "$HOME/.dotnet9" "$HOME/.nuget/NuGet"; \
  curl -sSL https://dot.net/v1/dotnet-install.sh | bash /dev/stdin --quality daily --channel 9.0 --install-dir "$HOME/.dotnet9"; \
  cat > "$HOME/.nuget/NuGet/NuGet.Config" <<EOF
<?xml version="1.0" encoding="utf-8"?>
<configuration>
<packageSources>
    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
    <add key="dotnet9" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet9/nuget/v3/index.json" />
</packageSources>
</configuration>
EOF

ENV DOTNET_NOLOGO=1
ENV DOTNET_EnableWriteXorExecute=0

RUN export dotnet9="$HOME/.dotnet9/dotnet"; \
  "$dotnet9" new webapiaot -n webapi1 && \
  "$dotnet9" publish webapi1 -o dist -c Release

FROM --platform=linux/arm/v7 ubuntu:latest
COPY --from=builder /root/webapi1/dist /app

CMD ["/app/webapi1"]
```

How to run it:

```sh
$ mkdir /tmp/arm-builder
$ cd /tmp/arm-builder
$ nano Dockerfile
# paste the contents posted above, save and close the editor.
$ docker build . -t armv7-nativeaot-webapi
```

>  > [builder 4/4] RUN export dotnet9="$HOME/.dotnet9/dotnet";   "$dotnet9" new webapiaot -n webapi1 &&   "$dotnet9" publish webapi1 -o dist -c Release:                                                                                               
10.40 The template "ASP.NET Core Web API (native AOT)" was created successfully.                                                                                                                                                                     
10.41                                                                                                                                                                                                                                                
10.41 Processing post-creation actions...                                                                                                                                                                                                            
10.41 Restoring /webapi1/webapi1.csproj:
15.71   Determining projects to restore...
31.18 Assertion failed: (dc->base.pc_next & 1) == 0 (../target/arm/tcg/translate.c: thumb_tr_translate_insn: 9428)
31.20 Aborted

**Comment by filipnavara** [comment]
> I ran into an assertion.

That’s assertion in QEMU. While I enjoy debugging things like these it’s not likely to get fixed any time soon :)

**Comment by am11** [comment]
> That’s assertion in QEMU.

Yup, from https://gitlab.com/qemu-project/qemu/-/blob/02e16ab9f4f19c4bdd17c51952d70e2ded74c6bf/target/arm/tcg/translate.c#L9429.

**Comment by am11** [comment]
@sonatique, meanwhile, we can use cross build option using host architecture builder and then test the binary with emulation (nativeaot binaries are not crashing under qemu).

Dockerfile:

```Dockerfile
FROM --platform=$BUILDPLATFORM ubuntu:latest AS builder

RUN apt update && apt install -y clang debootstrap curl

RUN mkdir -p "$HOME/.dotnet9" "$HOME/.nuget/NuGet"; \
  curl -sSL https://dot.net/v1/dotnet-install.sh | bash /dev/stdin --quality daily --channel 9.0 --install-dir "$HOME/.dotnet9"; \
  cat > "$HOME/.nuget/NuGet/NuGet.Config" <<EOF
<?xml version="1.0" encoding="utf-8"?>
<configuration>
<packageSources>
    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
    <add key="dotnet9" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet9/nuget/v3/index.json" />
</packageSources>
</configuration>
EOF

ENV DOTNET_NOLOGO=1
ENV ROOTFS_DIR=/crossrootfs/arm

RUN mkdir /dev/arm; \
  curl -sSL https://raw.githubusercontent.com/dotnet/arcade/main/eng/common/cross/arm/sources.list.bionic -o /dev/arm/sources.list.bionic; \
  curl -sSL https://raw.githubusercontent.com/dotnet/arcade/main/eng/common/cross/build-rootfs.sh |\
    bash /dev/stdin arm bionic llvm14

RUN export dotnet9="$HOME/.dotnet9/dotnet"; \
  "$dotnet9" new webapiaot -n webapi1 && \
  "$dotnet9" publish webapi1 -o dist -c Release -r linux-arm -p:SysRoot=/.tools/rootfs/arm
```

Usage:
```sh
$ mkdir /tmp/arm-builder
$ cd /tmp/arm-builder
$ nano Dockerfile
# paste the contents posted above, save and close the editor.

# build the container
$ docker build . -t armv7-nativeaot-webapi
# create a temporary instance and copy the artifacts locally
$ docker cp $(docker create --name webapi1 armv7-nativeaot-webapi):/dist/webapi1 .
$ docker rm webapi1

# test the binary by running it with arm container
$ docker run -e 'ASPNETCORE_URLS=http://*:8080' --rm -v$(pwd):/app --platform linux/arm/v7 \
    -p 8080:8080 ubuntu /app/webapi1

$ curl -s http://localhost:8080/todos | jq
[
  {
    "id": 1,
    "title": "Walk the dog",
    "dueBy": null,
    "isComplete": false
  },
  {
    "id": 2,
    "title": "Do the dishes",
    "dueBy": "2024-04-11",
    "isComplete": false
  },
  {
    "id": 3,
    "title": "Do the laundry",
    "dueBy": "2024-04-12",
    "isComplete": false
  },
  {
    "id": 4,
    "title": "Clean the bathroom",
    "dueBy": null,
    "isComplete": false
  },
  {
    "id": 5,
    "title": "Clean the car",
    "dueBy": "2024-04-13",
    "isComplete": false
  }
]
```

**Comment by am11** [comment]
Hey @sonatique, sorry for the confusion. I was testing these steps in a container and stitched together as a Dockefile, should had tested the final version. 😅 

I've now updated the docker and executed all steps in WSL, lets give it another try!

Changes:

* install lld in base layer: lld doesn't require separate cross package; it's multi-targetting. Specify `-p:LinkerFlavor=lld` in publish command.
  * it was either that or apt install binutils-arm. both are ok options (binuitls/bfd is slightly better at size optimization).
* switch to ubuntu jammy (v22.04) which has llvm15 which bionic (20.04) doesn't. It was either this or use `llvm10 lldb10` with bionic.
  * the benefit of using "old" distro is that it maximizes binary compatibility; rule is: older distro means older glibc, and the old the libc, the greater the compat (if you wanted to copy the binary on some device or distribute it for public e.g.).
    * Note: this only applies to libc version, the compiler toolchain rule is the opposite, the latest the better.
  * the default selection in build-rootfs.sh is 3.9 which is very old and not recommended for AOT. you can check the latest version of available toolchain using `docker run --rm --platform linux/arm/v7 ubuntu:jammy sh -c 'apt update && apt search llvm'` with codename of your choice (bionic etc. or versions as label `ubuntu:22.04`).
* install llvm in base layer: we need tools like llvm-objcopy. again, it was either that or install binuilts.

**Comment by sonatique** [comment]
@am11 : trying now. Thanks a lot!

**Comment by sonatique** [comment]
@am11 . I am still getting the same issue. I must be doing something wrong. 
Here is what I get after I docker built with your newer dockerfile. Everything went smooth up to the dotnet publish command:
```
/root/.dotnet9/dotnet publish -r linux-arm -v diag
/root/.dotnet9/sdk/9.0.100-preview.4.24215.10/MSBuild.dll -nologo --property:_IsPublishing=true -property:RuntimeIdentifier=linux-arm -property:_CommandLineDefinedRuntimeIdentifier=true -property:Configuration=Release -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,/root/.dotnet9/sdk/9.0.100-preview.4.24215.10/dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,/root/.dotnet9/sdk/9.0.100-preview.4.24215.10/dotnet.dll -maxcpucount -restore -target:Publish -tlp:default=auto -verbosity:m -verbosity:diag ./test-console-1.csproj
Restore complete (0.9s)
    Determining projects to restore...
    All projects are up-to-date for restore.
You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
  test-console-1 failed with 2 error(s) (1.5s) → bin/Release/net9.0/linux-arm/test-console-1.dll
    /usr/bin/ld.bfd: unrecognised emulation mode: armelf_linux_eabi
    Supported emulations: elf_x86_64 elf32_x86_64 elf_i386 elf_iamcu elf_l1om elf_k1om i386pep i386pe
    clang : error : linker command failed with exit code 1 (use -v to see invocation)
```
any idea by chance? Note that I am on a x64 machine. Since Windows ARM64 is not very common I didn't mention it earlier, but I see no arm* "supported emulations" in the list, I wonder why.

In case you wonder, here is my csproj file:
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net9.0</TargetFramework>
    <RootNamespace>test_console_1</RootNamespace>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
	<PublishTrimmed>true</PublishTrimmed>
	<PublishSingleFile>true</PublishSingleFile>
	<InvariantGlobalization>true</InvariantGlobalization>
	<PublishAot>true</PublishAot>
  </PropertyGroup>

</Project>

```
and I just have the default program.cs generated by `dotnet new console`

Thanks in advance!

**Comment by sonatique** [comment]
Thanks again @am11 ! 

Now that I better understand what I am doing and that I have a working setup, I am curious about one of the things you wrote earlier. You said you "installed lld, it was either that or apt install binutils-arm" and later you wrote the same about llvm. 

So you appear to have made efforts to avoid "binutils-arm", though you wrote "binuitls/bfd is slightly better at size optimization".

So I am wondering: what would be wrong in "simply" using binutils? I have to admit I tried for some time to install binutils and remove or replace `-p:LinkerFlavor=lld` and `-p:ObjCopy=llvm-objcopy` but to no avail, so it's probably not trivial (at least to me).

Do you have any pointer regarding how to use binutils? I wish I could compare binary produced by both lld/llvm and binutils.

Thanks in advance if you have time for this low priority question.

**Comment by am11** [comment]
gcc toolchain (gcc, binutils etc.) are architecture-specific, while llvm toolchain is multiarch. Meaning to cross-compile stuff, lld from llvm toolchain of host architecture will do the job, while gcc requires target arch specific package, e.g. `apt install -y binutils-arm-linux-gnueabi`.

See the previous attempt of using gcc toolchian in crosscompile: https://github.com/dotnet/runtime/issues/78559, the conclusion was it's bet to stick with llvm toolchain for cross-compilation. The difference in size is a few KBs and it's not meaningful in grand scheme of things.

**Comment by sonatique** [comment]
Hi @am11 , OK I see, thanks a lot!

**Comment by tunger** [comment]
Thanks a lot for the Dockerfile, @am11, and for asking the questions, @sonatique!

With the provided Dockerfile, we can utilize vscode [dev containers](https://code.visualstudio.com/docs/devcontainers/create-dev-container) to develop net9.0 apps in the IDE and compile it AOT for linux-arm. It works great!

Awesome work on the native AOT for linux-arm. It speeds up my application a lot.

Set it up like this:

`.devcontainer/devcontainer.json`
```json
{
  "build": { "dockerfile": "Dockerfile" },

  "customizations": {
    "vscode": {
      "extensions": ["ms-dotnettools.csdevkit"]
    }
  }
}
```

`.devcontainer/Dockerfile`
```dockerfile
FROM --platform=$BUILDPLATFORM ubuntu:latest AS builder

RUN apt update && apt install -y clang debootstrap curl lld llvm

RUN mkdir /dev/arm; \
  curl -sSL https://raw.githubusercontent.com/dotnet/arcade/main/eng/common/cross/arm/sources.list.jammy -o /dev/arm/sources.list.jammy; \
  curl -sSL https://raw.githubusercontent.com/dotnet/arcade/main/eng/common/cross/build-rootfs.sh |\
    bash /dev/stdin arm jammy llvm15 lldb15
	
RUN mkdir -p "$HOME/.dotnet" "$HOME/.nuget/NuGet";
RUN curl -sSL https://dot.net/v1/dotnet-install.sh | bash /dev/stdin --quality daily --channel 9.0;
RUN cat > "$HOME/.nuget/NuGet/NuGet.Config" <<EOF
<?xml version="1.0" encoding="utf-8"?>
<configuration>
<packageSources>
    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
    <add key="dotnet9" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet9/nuget/v3/index.json" />
</packageSources>
EOF

ENV DOTNET_NOLOGO=1
ENV PATH "$PATH:/root/.dotnet"
```

Then you can use vscode tasks to publish the app, like this:

`.vscode/tasks.json`
```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "command": "dotnet",
      "args": [
        "publish",
        "application.csproj",
        "-r",
        "linux-arm",
        "-c",
        "Release",
        "--self-contained",
        "true",
        "-o",
        "${workspaceFolder}/out",
        "-p:PublishSingleFile=false",
        "-p:EnableCompressionInSingleFile=true",
        "-p:PublishAot=true",
        "-p:LinkerFlavor=lld",
        "-p:ObjCopy=llvm-objcopy",
        "-p:SysRoot=\"/.tools/rootfs/arm\""
      ],
      "options": {
        "cwd": "${workspaceFolder}"
      },
      "group": "build",
      "label": "dotnet9 publish release AOT"
    }
  ]
}
```

**Comment by am11** [comment]
@tunger, very nice! Thanks for sharing. After #101213 is merged, we can use the same mechanism for `linux-musl-arm` targeting Alpine Linux. Basically, in the Dockerfile:

```diff
- bash /dev/stdin arm jammy llvm15 lldb15
+ bash /dev/stdin arm alpine llvm15 lldb15
```
and in tasks.json:
```diff
        "-r",
-        "linux-arm",
+        "linux-musl-arm",
```

Once .NET 9 is shipped, we would be able to use the prebuilt official docker images which will exempt setting up cross environment.

One reason of recommending official images is slightly(⚓)  important because the more people start using this kind of experimental solution, the more confusion it's going to cause. For instance; this dockerfile (for linux-arm and not for linux-musl-arm) requires 'nested virtualization' support for chroot (fakechroot has some issues so `fakeroot fakechroot debootstrap ... -variant=fakechroot` also ends up requiring this the real chroot.. which - according to google results - is a known issue actively being investigated in Debian world). Nested virtualization needs to "at least" EL0 (aka `Non-hardware Assisted Nested Virtualization`; the [kind](https://apple.stackexchange.com/a/467030) which is available on M1 macs, which does not support hardware assisted virtualization). Alpine's crossbuild setup, OTOH, doesn't require chroot etc., and it is lightening faster than debootstrap (yet to find any package manager faster than Alpine's `apk(1)` 😁).

I tested this docker on a few public CI systems, here is the support situation:
* appveyor: no EL0 support, so debootsrap throws fatal exit code at chroot step
* cirrus-ci: same as appveyor
* GitHub Actions: only Linux supports it. So it's best to publish your base layer to something like [Docker or GitHub Package Registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry) to avoid the surprises.

⚓ It is only about "building" the image, dotnet-publish does not use chroot.

**Comment by dabbinavo** [comment]
Still the same failure (`docker build`, as well directly inside WSL) after executing your suggested commands inside WSL
```sh
sudo update-binfmts --display
sudo update-binfmts --enbale
```

**Comment by filipnavara** [comment]
`enbale` is a typo, right? `update-binfmts --display` is supposed to list all the various QEMU user configurations, does it?

**Comment by dabbinavo** [comment]
Yes, it was a typo and the display switch outputs:
```
~$ update-binfmts --display
llvm-14-runtime.binfmt (enabled):
     package = llvm-14-runtime
        type = magic
      offset = 0
       magic = BC
        mask =
 interpreter = /usr/bin/lli-14
    detector =
python3.10 (enabled):
     package = python3.10
        type = magic
      offset = 0
       magic = \x6f\x0d\x0d\x0a
        mask =
 interpreter = /usr/bin/python3.10
    detector =
```

**Comment by filipnavara** [comment]
That doesn't list any of the QEMU user packages. There would be an entry similar to:
```
qemu-arm (disabled):
     package = qemu-arm
        type = magic
      offset = 0
       magic = \x7fELF\x01\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x02\x00\x28\x00
        mask = \xff\xff\xff\xff\xff\xff\xff\x00\xff\xff\xff\xff\xff\xff\xff\xff\xfe\xff\xff\xff
 interpreter = /usr/bin/qemu-arm-static
    detector =
```

That means the QEMU user emulators are not installed properly and thus the chroot binaries in the debootstrap process cannot execute. I don't remember anymore how I fixed this but hopefully this points in the right direction to Google and fix the problem.

**Comment by am11** [comment]
I was using this docker https://github.com/am11/CrossRepoCITesting/blob/master/linux-arm-aot/Dockerfile and this workflow https://github.com/am11/CrossRepoCITesting/blob/master/.github/workflows/docker-naot-arm32.yml. I found that running `docker run --privileged --rm tonistiigi/binfmt --install all` before hand (as I did in the workflow yml) made things whole lot easier. Also, using that command fixed Appveyor CI and Cirrus-CI builds.

**Comment by dabbinavo** [comment]
I was able to compile an app for an embedded Debian 11 (bullseye) system by following these setup steps on my build machine (Azure Devops Services, Microsoft Hosted Agent, vmImage: ubuntu-20.04):

```sh
sudo apt install -y clang debootstrap curl lld llvm qemu-user-static binfmt-support
mkdir -p "$HOME/.dotnet9" "$HOME/.nuget/NuGet"
curl -sSL https://dot.net/v1/dotnet-install.sh | bash /dev/stdin --quality daily --channel 9.0 --install-dir "$HOME/.dotnet9"
cat > "$HOME/.nuget/NuGet/NuGet.Config" <<EOF
<?xml version="1.0" encoding="utf-8"?>
<configuration>
<packageSources>
    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
    <add key="dotnet9" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet9/nuget/v3/index.json" />
</packageSources>
</configuration>
EOF

export DOTNET_NOLOGO=1
export ROOTFS_DIR=/crossrootfs/arm

sudo mkdir /dev/arm
curl -sSL https://raw.githubusercontent.com/dotnet/arcade/main/eng/common/cross/arm/sources.list.focal -o /dev/arm/sources.list.focal
curl -sSL https://raw.githubusercontent.com/dotnet/arcade/main/eng/common/cross/build-rootfs.sh | sudo -E bash /dev/stdin arm focal llvm10 lldb10
```

Then running the cross compile build with:
```sh
/home/vsts/.dotnet9/dotnet publish /path/to/csharp-project --property:RuntimeIdentifier="linux-arm" --property:TargetFramework="net9.0" --property:PublishAot="true" --property:Configuration="Release" --property:LinkerFlavor="lld" --property:ObjCopy="llvm-objcopy" --property:SysRoot="/crossrootfs/arm" -o /desired/output/directory
```

As mentioned in [this comment](https://github.com/dotnet/runtime/issues/97729#issuecomment-2058764475), I had to use Ubuntu 20.04 and `focal` rootfs with `llvm10` and `lldb10` in order to run it on the bullseye system with "old" glibc.

**Comment by dabbinavo** [comment]
Another question from my side: Does anyone know if there will be "official" support for natively cross-compiling to linux-arm for .NET 9 and maybe also newer versions? The question comes up, as we are currently evaluating if it will be safe to use this functionality in an consumer product.

I mean will .NET 9 be internally testet against NativeAOT support for linux-arm and will it be maintained and bug-fixed after .NET 9 is released?

Unfortunately, I could not find any announcement or anything about this.

**Comment by am11** [comment]
As it happened, there already are tags for arm32 with .NET 9 SDK https://hub.docker.com/_/microsoft-dotnet-nightly-sdk/.  👌😎

```sh
# run a throw-away-after-use (--rm) container interactively for linux/arm/v7,
# while mounting the current-working-directory to /myapp 
$ docker run --rm --platform linux/arm/v7 -v$(pwd):/myapp -w /myapp -it \
      mcr.microsoft.com/dotnet/nightly/sdk:9.0-preview

# inside the container
$ uname -a
Linux ea9e301095ba 6.6.26-linuxkit #1 SMP Sat Apr 27 04:13:19 UTC 2024 armv7l GNU/Linux

$ dotnet --info
Linux ea9e301095ba 6.6.26-linuxkit #1 SMP Sat Apr 27 04:13:19 UTC 2024 armv7l GNU/Linux
root@ea9e301095ba:/# dotnet --info
.NET SDK:
 Version:           9.0.100-preview.4.24266.28
 Commit:            75a08fda5c
 Workload version:  9.0.100-manifests.2c9affbd
 MSBuild version:   17.11.0-preview-24225-01+bd0b1e466

Runtime Environment:
 OS Name:     debian
 OS Version:  12
 OS Platform: Linux
 RID:         linux-arm
 Base Path:   /usr/share/dotnet/sdk/9.0.100-preview.4.24266.28/

.NET workloads installed:
There are no installed workloads to display.

Host:
  Version:      9.0.0-preview.4.24260.3
  Architecture: arm
  Commit:       2270e3185f

.NET SDKs installed:
  9.0.100-preview.4.24266.28 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 9.0.0-preview.4.24260.3 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 9.0.0-preview.4.24260.3 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

On macOS arm64, I run into qemu assertion https://github.com/dotnet/runtime/issues/97729#issuecomment-2050507728.

If you are on the x64 host with docker installed, you can use one of these tags with prerequisites for cross compilation: https://github.com/dotnet/versions/blob/main/build-info/docker/image-info.dotnet-dotnet-buildtools-prereqs-docker-main.json

e.g.

```sh
$ docker run -e ROOTFS_DIR=/crossrootfs/arm --rm -it mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-cross-arm
# install dotnet in it
```
You can also create a Dockerfile to make it ready.
e.g.
```dockerfile
FROM mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-cross-arm

# install dotnet (dotnet-script install)

# now warm up NativeAOT so ilc packages are ready to use
RUN dotnet new webapiaot -n warmupapp && dotnet publish --project warmupapp && rm -rf warmupapp
```

then build this image and tag it `docker build . -t my-dotnet9-linux-arm-builder`. Usage `dotnet run --rm -v$(pwd):/myapp -w /myapp my-dotnet9-linux-arm-builder  dotnet publish -c Release -o dist`.

However, if you are on non-x64 machine like arm64, or you wanted the bleeding-edge daily build then you can build the builder image from scratch as discussed above.

**Comment by am11** [comment]
> Using a newer docker image fixed this issue.

Yea, use it successor `azurelinux-3.0-cross-arm-{alpine-}net9.0` as CBL mariner is frozen.

**Comment by am11** [comment]
> Using a newer docker image fixed this issue.

Yea, use its successor `azurelinux-3.0-cross-arm-{alpine-}net9.0` as CBL mariner is frozen.

**Comment by jkotas** [comment]
linux-arm is officially supported for NativeAOT starting with .NET 9: https://learn.microsoft.com/dotnet/core/deploying/native-aot/?tabs=windows%2Cnet9plus#platformarchitecture-restrictions

Thank you all for making this happen!
